agent/grpc-external/metadata.go

```diff
@@ -4,11 +4,28 @@ import (
 	"context"
 	"fmt"
 
-	"github.com/hashicorp/consul/agent/structs"
 	"github.com/mitchellh/mapstructure"
 	"google.golang.org/grpc/metadata"
+
+	"github.com/hashicorp/consul/agent/structs"
 )
 
+const metadataKeyToken = "x-consul-token"
+
+// TokenFromContext returns the ACL token in the gRPC metadata attached to the
+// given context.
+func TokenFromContext(ctx context.Context) string {
```

Comment:
Reverting a deletion from https://github.com/hashicorp/consul/pull/14724/files#diff-b4bf50db93d52f639b211a36e1e38f0e6d47a166ba03842d84c0a0251811e42a

Using `mapstructure` to extract the token seemed a bit inefficient for xDS

Reply:
Relatively speaking mapstructure is expensive but it's still in the realm of nanoseconds so certainly not too expensive.
```go
func BenchmarkTokenFromContext(b *testing.B) {
	md := metadata.MD{}
	md.Set(metadataKeyToken, "9df2d1a4-2d07-414e-8ead-6053f56ed2eb")
	ctx := metadata.NewOutgoingContext(context.Background(), md)

	var token string
	for i := 0; i < b.N; i++ {
		token = TokenFromContext(ctx)
	}
	_ = token
}

func BenchmarkQueryOptionsFromContext(b *testing.B) {
	md := metadata.MD{}
	md.Set(metadataKeyToken, "9df2d1a4-2d07-414e-8ead-6053f56ed2eb")
	ctx := metadata.NewOutgoingContext(context.Background(), md)

	var token string
	for i := 0; i < b.N; i++ {
		o, _ := QueryOptionsFromContext(ctx)
		token = o.Token
	}
	_ = token
}
```

```
BenchmarkQueryOptionsFromContext-16    	18966411	        61.19 ns/op	      96 B/op	       1 allocs/op
BenchmarkTokenFromContext-16    	165877292	         7.099 ns/op	       0 B/op	       0 allocs/op
```

I don't mind reusing QueryOptionsFromContext